doc/man/hdfsee.1
.\"
.\" hdfsee man page
.\"
.\" Copyright © 2012-2020 Deutsches Elektronen-Synchrotron DESY,
.\"                       a research centre of the Helmholtz Association.
.\"
.\" Part of CrystFEL - crystallography with a FEL
.\"

.TH HDFSEE 1
.SH NAME
hdfsee \- HDF5 image viewer
.SH SYNOPSIS
.PP
.B hdfsee \fIimage.h5\fR [\fIoptions\fR] \fB...\fR
.PP
\fBhdfsee --help\fR

.SH DESCRIPTION
hdfsee is a simple image viewer for images stored in HDF5 files.
.PP
Most of the options that can be set on the command line can be changed via the user interface once the program is running.  The command line options are to help when running hdfsee from a script.

.SH OPTIONS
.PD 0
.IP "\fB-p\fR \fIfilename\fR"
.IP \fB--peak-overlay=\fR\fIfilename\fR
.PD
Peak locations will be read from \fIfilename\fR and displayed on the image.  The peak location file can use the format used in CrystFEL stream files for lists of integrated reflections, i.e. including the Miller indices.  If a line cannot be read using this format, it will be assumed to be a simple two-column list of fast scan and slow scan coordinates.  Any text beyond the second column will be ignored.  If the line cannot be read in either format, it will be ignored.

.PD 0
.IP \fB--ring-size=\fR\fIradius\fR
.PD
Set the radius of the rings used for displaying peak locations.  The radius is given in pixels on the screen, i.e. the circles are drawn \fIafter\fR binning the image data.

.PD 0
.IP "\fB-b\fR \fIb\fR"
.IP \fB--binning=\fR\fIb\fR
.PD
Show the image after binning down by a factor of \fIb\fR.

.PD 0
.IP "\fB-e\fR \fIpath\fR"
.IP \fB--image=\fR\fIpath\fR
.PD
Get the image data to display from \fIpath\fR inside the HDF5 file.  Example: \fI/data/rawdata\fR.
.IP
If no geometry file is provided (see \fB-g\fR) and this option is not used, hdfsee will display the first two-dimensional dataset it finds in the file with both dimensions greater than 64.  If a geometry file is provided, the data layout description from the file determines which image is displayed (see \fB man crystfel_geometry\fR).
.IP
If \fB-e\fR is used in combination with \fB-g\fR, hdfsee will attempt to show the specified dataset with the geometry applied.  The data layout description from the geometry file will be ignored, the data block to be used as data source for all panels will be set to \fIpath\fR and it will be assumed to be 2-dimensional.

.PD 0
.IP \fB--event=\fIID\fR
.PD
Show the event identified by \fIID\fR.  Use this when you have a multi-event HDF5 file and the corresponding geometry file.

.PD 0
.IP "\fB-g\fR \fIfilename\fR"
.IP \fB--geometry=\fR\fIfilename\fR
.PD
Display the image data according to the geometry description in \fIfilename\fR.  Out of plane components of the geometry will be ignored.  This is required to display resolution rings.

.PD 0
.IP "\fB-i\fR \fIn\fR"
.IP \fB--int-boost=\fR\fIn\fR
.PD
Multiply the intensity in the image by \fIn\fR before displaying.  With n=1, the top of the colour scale will represent the maximum pixel intensity found in the image.

.PD 0
.IP \fB--show-rings\fR
.PD
Show resolution rings on the image at 1 Angstrom intervals.  You must provide both a geometry and a beam file to use this option.

.PD 0
.IP \fB--simple-rings=\fR\fIradii\fR
.PD
Show rings on the image with the radii specified.  \fIradii\fR can be a comma-deliminated list of several values, for example \fI100,200\fR.  The radii have units of pixels on the detector before bininng.  You must provide a geometry file to use this option.

.PD 0
.IP "\fB-c\fR \fIscale\fR"
.IP \fB--colscale=\fR\fIscale\fR
.PD
Use \fIscale\fR as the colour scale.  Possible scales are: \fBmono\fR, \fBinvmono\fR and \fBcolour\fR.

.PD 0
.IP \fB--median-filter=\fR\fIn\fR
.PD
Apply a median filter with box "radius" \fIn\fR to the image.  Each pixel will be set to the median of the values from a \fI(n+1)\fRx\fI(n+1)\fR square centered on the pixel.  If you also use \fB--noise-filter\fR, the median filter will be applied first.

.PD 0
.IP \fB--filter-noise\fR
.PD
Apply a noise filter to the image with checks 3x3 squares of pixels and sets all of them to zero if any of the nine pixels have a negative value.  If you also use \fB--median-filter\fR, the median filter will be applied first.

.PD 0
.IP "\fB-o \fIcoll\fR"
.IP \fB--rigid-groups=\fIcoll\fR
.PD
Use \fIcoll\fR as the rigid group collection for calibration mode.

.SH CALIBRATION MODE
Calibration mode allows you to visually adjust the locations of panels.  To enter calibration mode, select Tools->Calibration Mode from the menu.  The currently selected panel will be bordered in white.  Press + or - to move to the next or previous panel (as listed in the geometry file).  Use the arrow keys to move the current panel.  Press 'f' to hide or restore the white border.  Press 's' to save the geometry file with your modifications.  Press 'g' to toggle between moving individual panels, rigid groups (if any are defined in the geometry file) and moving all panels together.
Most of these actions can also be accessed from the Calibration menu, which becomes aptive when calibration mode is toggled once

.SH EVENT NAVIGATION
When multi-event files are opened, the Events menu in the menubar becomes active, and some event navigation tools become available. The title bar shows, in addition to the file name, the name of the event currently displayed (See \fBman crystfel_geometry\fR and \fBman indexamajig\fR for a description of event naming). Press 'n' to move to the next event in the file, and 'p' to move to the previous one'. Press 'e' to jump to a specific event, by providing an event name (Use the \fBlist_events\fR program to get a list of the events included in a file). Press 'r' to jump to a random event. These actions are also accessible from the Events menu in the menubar.

.SH AUTHOR
This page was written by Thomas White and Valerio Mariani.

.SH REPORTING BUGS
Report bugs to <taw@physics.org>, or visit <http://www.desy.de/~twhite/crystfel>.

.SH COPYRIGHT AND DISCLAIMER
Copyright © 2012-2020 Deutsches Elektronen-Synchrotron DESY, a research centre of the Helmholtz Association.
.P
hdfsee, and this manual, are part of CrystFEL.
.P
CrystFEL is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.
.P
CrystFEL is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
.P
You should have received a copy of the GNU General Public License along with CrystFEL.  If not, see <http://www.gnu.org/licenses/>.

.SH SEE ALSO
.BR crystfel (7),
.BR pattern_sim (1),
.BR crystfel_geometry (5).


NEWS
There is no NEWS.  Check the version control repository...


src/hdfsee.c
/*
 * hdfsee.c
 *
 * Quick yet non-crappy HDF viewer
 *
 * Copyright © 2012-2020 Deutsches Elektronen-Synchrotron DESY,
 *                       a research centre of the Helmholtz Association.
 * Copyright © 2012 Richard Kirian
 *
 * Authors:
 *   2009-2014 Thomas White <taw@physics.org>
 *   2014      Valerio Mariani
 *   2012      Richard Kirian
 *
 * This file is part of CrystFEL.
 *
 * CrystFEL is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * CrystFEL is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with CrystFEL.  If not, see <http://www.gnu.org/licenses/>.
 *
 */


#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <gtk/gtk.h>
#include <getopt.h>

#include "dw-hdfsee.h"
#include "utils.h"
#include "render.h"


/* Global program state */
DisplayWindow *main_window_list[64];
size_t main_n_windows = 0;


static void show_help(const char *s)
{
	printf("Syntax: %s [options] image.h5\n\n", s);
	printf(
"Quick HDF5 image viewer.\n"
"\n"
"  -h, --help                       Display this help message.\n"
"      --version                    Print CrystFEL version number and exit.\n"
"\n"
"  -p, --peak-overlay=<filename>    Draw circles in positions listed in file.\n"
"      --ring-size=<n>              Set the size for those circles.\n"
"  -i, --int-boost=<n>              Multiply intensity by <n>.\n"
"  -b, --binning=<n>                Set display binning to <n>.\n"
"      --filter-noise               Apply an aggressive noise filter to the\n"
"                                    image data.\n"
"      --median-filter=<n>          Apply a median filter to the image data.\n"
"      --calibration-mode           Start in calibration mode\n"
"      --show-rings                 Overlay rings that indicate resolution.\n"
"      --simple-rings=XX,YY,...     Overlay rings at specified radii XX, YY, ...\n"
"                                    in pixel units.\n"
"  -c, --colscale=<scale>           Use the given colour scale.  Choose from:\n"
"                                    mono    : Greyscale, black is zero.\n"
"                                    invmono : Greyscale, white is zero.\n"
"                                    colour  : Colour scale:\n"
"                                               black-blue-pink-red-orange-\n"
"                                               -yellow-white.\n"
"  -e, --image=<element>            Start up displaying this image from the\n"
"                                    HDF5 file. When this option is used,\n"
"                                    information about the data layout\n"
"                                    from the geometry file is ignored (See\n"
"                                    manual page).\n"
"                                    Example: /data/data0.\n"
"      --event=<event code>         Event to show from multi-event file.\n"
"  -g, --geometry=<filename>        Use geometry from file for display.\n"
"                                   (When this option is used, the value of\n"
"                                    of the -e parameter is ignored)\n"
"  -o, --rigid-groups=<coll>        Use rigid group collection <coll>.\n"
"\n");
}


/* Called to notify that an image display window has been closed */
void hdfsee_window_closed(DisplayWindow *dw)
{
	size_t i;

	for ( i=0; i<main_n_windows; i++ ) {

		if ( main_window_list[i] == dw ) {

			size_t j;

			for ( j=i+1; j<main_n_windows; j++ ) {
				main_window_list[j] = main_window_list[j+1];
			}

		}

	}

	main_n_windows--;

	if ( main_n_windows == 0 ) gtk_main_quit();

}


int main(int argc, char *argv[])
{
	int c;
	size_t i;
	int nfiles;
	char *peaks = NULL;
	char *geom_filename = NULL;
	double boost = 1.0;
	int binning = 2;
	int config_noisefilter = 0;
	int config_showrings = 0;
	int config_calibmode =0;
	int colscale = SCALE_COLOUR;
	char *cscale = NULL;
	char *element = NULL;
	char *event = NULL;
	char *rgcoll_name = NULL;
	double ring_size = 5.0;
	char *reslist = NULL;
	double ring_radii[128];
	int n_rings = -1;
	int median_filter = 0;
	struct detector *det_geom = NULL;
	struct beam_params cbeam;
	struct beam_params *beam = NULL;

	/* Long options */
	const struct option longopts[] = {
		{"help",               0, NULL,               'h'},
		{"version",            0, NULL,                4 },
		{"peak-overlay",       1, NULL,               'p'},
		{"int-boost",          1, NULL,               'i'},
		{"binning",            1, NULL,               'b'},
		{"filter-noise",       0, &config_noisefilter, 1},
		{"colscale",           1, NULL,               'c'},
		{"image",              1, NULL,               'e'},
		{"geometry",           1, NULL,               'g'},
		{"show-rings",         0, &config_showrings,   1},
		{"ring-size",          1, NULL,                2},
		{"simple-rings",       1, NULL,               'r'},
		{"median-filter",      1, NULL,                3},
		{"calibration-mode",   0, &config_calibmode,   1},
		{"event",              1, NULL,                5},
		{"rigid-groups",       1, NULL,               'o'},
		{0, 0, NULL, 0}
	};

	/* Default beam parameters */
	cbeam.photon_energy = 0.0;
	cbeam.photon_energy_from = NULL;

	/* This isn't great, but necessary to make the command-line UI and file
	 * formats consistent with the other programs, which all use the C
	 * locale.  Better would be to have all the programs call
	 * setlocale(LC_ALL, "") and use the C locale temporarily when reading
	 * or writing a stream, reflection file, geometry file etc. */
	gtk_disable_setlocale();

	gtk_init(&argc, &argv);

	/* Short options */
	while ((c = getopt_long(argc, argv, "hp:b:i:c:e:g:2:r:m:o:",
	                        longopts, NULL)) != -1) {

		char *test;

		switch (c) {

			case 'h' :
			show_help(argv[0]);
			return 0;

			case 4 :
			printf("CrystFEL: " CRYSTFEL_VERSIONSTRING "\n");
			printf(CRYSTFEL_BOILERPLATE"\n");
			return 0;

			case 'p' :
			peaks = strdup(optarg);
			break;

			case 'i' :
			boost = atof(optarg);
			if ( boost <= 0 ) {
				ERROR("Intensity boost must be a positive"
				      " number.\n");
				return 1;
			}
			break;

			case 'b' :
			binning = atoi(optarg);
			if ( binning < 1 ) {
				ERROR("Binning must be a positive integer.\n");
				return 1;
			}
			break;

			case 'c' :
			cscale = strdup(optarg);
			break;

			case 'e' :
			element = strdup(optarg);
			break;

			case 'g' :
			geom_filename = strdup(optarg);
			det_geom = get_detector_geometry(geom_filename, &cbeam);
			if ( det_geom == NULL ) {
				ERROR("Failed to read detector geometry "
				      "from '%s'\n", optarg);
				return 1;
			}
			beam = &cbeam;
			break;

			case 'o' :
			rgcoll_name = strdup(optarg);
			break;

			case 2 :
			ring_size = strtod(optarg, &test);
			if ( test == optarg ) {
				ERROR("Ring size must be numerical.\n");
				return 1;
			}
			break;

			case 3 :
			median_filter = atoi(optarg);
			break;

			case 'r' :
			config_showrings = 1;
			reslist = strdup(optarg);
			int nchar = strlen(reslist);
			char thisvalue[128];
			int i;
			int j = 0;
			n_rings = 0;
			for ( i=0; i<=nchar; i++ ) {
				if ( ( reslist[i] != ',' )
				  && ( reslist[i] != '\0' ) )
				{
					thisvalue[j] = reslist[i];
					j++;
				} else {
					thisvalue[j] = '\0';
					ring_radii[n_rings] = atof(thisvalue);
					n_rings++;
					j = 0;
				}
			}
			break;

			case 5 :
			event = strdup(optarg);
			break;

			case 0 :
			break;

			case '?' :
			break;

			default :
			ERROR("Unhandled option '%c'\n", c);
			break;
		}

	}

	nfiles = argc-optind;

	if ( nfiles < 1 ) {
		ERROR("You need to give me a file to open!\n");
		return -1;
	}

	if ( (element != NULL) && (event != NULL) ) {
		ERROR("The options --event and --element are "
		      "mutually exclusive\n");
		return 1;
	}

	if ( event != NULL && geom_filename == NULL) {
		ERROR("The '--event' option requires geometry file\n");
		return 1;
	}

	if ( cscale == NULL ) cscale = strdup("colour");
	if ( strcmp(cscale, "mono") == 0 ) {
		colscale = SCALE_MONO;
	} else if ( strcmp(cscale, "invmono") == 0 ) {
		colscale = SCALE_INVMONO;
	} else if ( strcmp(cscale, "colour") == 0 ) {
		colscale = SCALE_COLOUR;
	} else if ( strcmp(cscale, "color") == 0 ) {
		colscale = SCALE_COLOUR;
	} else {
		ERROR("Unrecognised colour scale '%s'\n", cscale);
		return 1;
	}
	free(cscale);

	for ( i=0; i<nfiles; i++ ) {
		main_window_list[i] = displaywindow_open(argv[optind+i],
		                                         geom_filename,
		                                         peaks, boost, binning,
		                                         config_noisefilter,
		                                         config_calibmode,
		                                         colscale, element,
		                                         event, det_geom, beam,
		                                         rgcoll_name,
		                                         config_showrings,
		                                         ring_radii,
		                                         n_rings,
		                                         ring_size,
		                                         median_filter);
		if ( main_window_list[i] != NULL ) main_n_windows++;
	}

	if ( main_n_windows == 0 ) return 0;
	gtk_main();

	return 0;
}
